### DISCUSSION TITLE
Is it possible to see the stacktrace on reverts inside instantiated contracts?

### DISCUSSION THREAD

**Discussion by JustCallMeDavid** [NomicFoundation/hardhat#3552] [Troubleshooting] [answered]
I am trying to debug an error inside a chain of contract callbacks. There is one contract I deploy myself (`MyContract`) and the others are instantiated using their interfaces from real contract addresses on the local mainnet fork I am running, e.g.,
`IMainnetContract mainnetContract = IMainnetContract(0x...);`.

The problem I am having is that when I get an error, the stacktrace only recognizes and shows functions inside of my own contract, such as for instance:
```
  Error: Transaction reverted without a reason string
      at <UnrecognizedContract>.<unknown> (0x1..)
      at <UnrecognizedContract>.<unknown> (0x2..)
      at <UnrecognizedContract>.<unknown> (0x1..)
      at MyContract.MyFunction (../MyContract.sol:246)
      at <UnrecognizedContract>.<unknown> (0x3..)
      at MyContract.MyOtherFunction (../MyContract.sol:332)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
      at HardhatNode._mineBlockWithPendingTxs (..\node_modules\hardhat\src\internal\hardhat-network\provider\node.ts
:1815:23)
      at HardhatNode.mineBlock (..\node_modules\hardhat\src\internal\hardhat-network\provider\node.ts:504:16)       
      at EthModule._sendTransactionAndReturnHash ..\node_modules\hardhat\src\internal\hardhat-network\provider\modu
les\eth.ts:1522:18)
      at HardhatNetworkProvider._sendWithLogging ..\node_modules\hardhat\src\internal\hardhat-network\provider\prov
ider.ts:139:22)
      at HardhatNetworkProvider.request (..\node_modules\hardhat\src\internal\hardhat-network\provider\provider.ts:1
16:18)
      at JsonRpcHandler._handleRequest (..node_modules\hardhat\src\internal\hardhat-network\jsonrpc\handler.ts:188:
20)
```

My question is, whether there would be any way to make Hardhat recognize all the contracts in the stack trace? I assume they are not recognized because they are already deployed on mainnet and thus were not previously compiled with hardhat. Is there any way to work around this?

The only solution I can come up with would be to deploy copies of them and then call these copies, but this is not ideal since replicating their exact mainnet state is nontrivial.

I would be grateful for any suggestions on how to achieve a full stack trace with Hardhat or for any general ideas that may help.

**Comment by alcuadrado** [comment, marked as answer]
You can also download the sources from etherscan or an equivalent explorer, add them to your `contracts/` and it should work.

Please report back if it does or not?

**Comment by JustCallMeDavid** [comment]
Thank you @alcuadrado, indeed this works. Was not aware Hardhat would be able to recognize the trace then, amazingly it does 🙏

**Reply by JustCallMeDavid** [reply]
Could it be this does not yet handle recursive resolution? The contracts I instantiate directly are recognized but contracts instantiated in the same way by other instantiated contracts are not. I have all of them in the `/contracts` directory.

**Comment by JustCallMeDavid** [comment]
Sorry to append to a solved discussion, but would you @alcuadrado happen to know if there was some Hardhat magic that would allow inserting Hardhat's `console.log` statements inside of instantiated smart contracts?

I tried doing that by adding logs to the sources, but that led to the contract no longer being recognized in the stack trace (probably due to the resulting change in its bytecode?) and no logs were showing.

**Reply by fvictorio** [reply]
No, that's not possible right now. One possible, but involved, workaround would be something like this:

1. You modify the contract with the extra `console.log` calls
2. You compile it and get the `deployedBytecode` (not the deployment bytecode)
3. You use [`setCode`](https://hardhat.org/hardhat-network-helpers/docs/reference#setcode(address,-code)) to change the bytecode of the already deployed contract

**Reply by JustCallMeDavid** [reply]
Thank you @fvictorio, will keep that solution in mind if I run on issues otherwise not solvable.

Since you said it is not possible 'right now', is it maybe something that is planned in the future?

**Reply by fvictorio** [reply]
Yes, I think being able to do something like this would be super useful. I have no idea which basic features would enable it, but some sort of hot-reloading of already deployed contract's bytecode during debugging would be ideal, similar to what you can do with chrome's devtools in javascript files. Obviously we don't have an ETA for this :sweat_smile:

**Reply by JustCallMeDavid** [reply]
I agree, it would be a great feature to have since the logging utility of Hardhat is a lifesaver 😃 Thank you for your help @fvictorio
